File: workshop-exercises/02-clean-code/csharp/Application/Services/DiscountService.cs
using Workshop.Application.Interfaces;

namespace Workshop.Application.Services
{
    public class DiscountService : IDiscountService
    {
        public double CalculateDiscount(double amount, string customerType)
        {
            return PerformDiscountCalculation(amount, customerType);
        }

        private double PerformDiscountCalculation(double amount, string customerType)
        {
            var discountPercent = GetDiscountPercentage(customerType);
            return ApplyDiscountPercentage(amount, discountPercent);
        }

        private double GetDiscountPercentage(string customerType)
        {
            return DetermineDiscountRate(customerType);
        }

        private double DetermineDiscountRate(string customerType)
        {
            switch (customerType)
            {
                case "premium":
                    return GetPremiumDiscount();
                case "regular":
                    return GetRegularDiscount();
                default:
                    return GetDefaultDiscount();
            }
        }

        private double GetPremiumDiscount()
        {
            return 0.15; // 15%
        }

        private double GetRegularDiscount()
        {
            return 0.05; // 5%
        }

        private double GetDefaultDiscount()
        {
            return 0.0; // 0%
        }

        private double ApplyDiscountPercentage(double amount, double percentage)
        {
            return CalculateDiscountedAmount(amount, percentage);
        }

        private double CalculateDiscountedAmount(double amount, double percentage)
        {
            var discount = ComputeDiscountValue(amount, percentage);
            return SubtractDiscount(amount, discount);
        }

        private double ComputeDiscountValue(double amount, double percentage)
        {
            return amount * percentage;
        }

        private double SubtractDiscount(double amount, double discount)
        {
            return amount - discount;
        }
    }
}

File: workshop-exercises/02-clean-code/go/application/discount_service.go
package application

// =============================================================================
// DISCOUNT SERVICE
// =============================================================================

type DiscountService struct{}

func NewDiscountService() DiscountServiceInterface {
	return &DiscountService{}
}

func (d *DiscountService) CalculateDiscount(amount float64, customerType string) (float64, error) {
	return d.performDiscountCalculation(amount, customerType)
}

func (d *DiscountService) performDiscountCalculation(amount float64, customerType string) (float64, error) {
	discountPercent := d.getDiscountPercentage(customerType)
	return d.applyDiscountPercentage(amount, discountPercent), nil
}

func (d *DiscountService) getDiscountPercentage(customerType string) float64 {
	return d.determineDiscountRate(customerType)
}

func (d *DiscountService) determineDiscountRate(customerType string) float64 {
	switch customerType {
	case "premium":
		return d.getPremiumDiscount()
	case "regular":
		return d.getRegularDiscount()
	default:
		return d.getDefaultDiscount()
	}
}

func (d *DiscountService) getPremiumDiscount() float64 {
	return 0.15 // 15%
}

func (d *DiscountService) getRegularDiscount() float64 {
	return 0.05 // 5%
}

func (d *DiscountService) getDefaultDiscount() float64 {
	return 0.0 // 0%
}

func (d *DiscountService) applyDiscountPercentage(amount float64, percentage float64) float64 {
	return d.calculateDiscountedAmount(amount, percentage)
}

func (d *DiscountService) calculateDiscountedAmount(amount float64, percentage float64) float64 {
	discount := d.computeDiscountValue(amount, percentage)
	return d.subtractDiscount(amount, discount)
}

func (d *DiscountService) computeDiscountValue(amount float64, percentage float64) float64 {
	return amount * percentage
}

func (d *DiscountService) subtractDiscount(amount float64, discount float64) float64 {
	return amount - discount
}